Add option for "allow-namespace-imports" to noDuplicateImports rule

EdwardDrapkin opened the pull request:
#### PR checklist

- [X] Addresses an existing issue: #3418 

#### Overview of change:
Adds an option "allow-namespace-imports" that can be passed to the no duplicate imports rule that will make it skip over lines containing namespace imports.

This option will fix the following fairly common pattern that currently errors the rule:

```
import * as React from 'react';
import { NamedExoticComponent } from 'react';
```

With the option set, the following error is still caught:

```
import * as React from 'react';
import { NamedExoticComponent } from 'react';
import { ExoticComponent } from 'react';
```

When the new option is enabled, wildcard imports will be allowed on a separate line from named imports, but the rule will still fail for multiple lines of named imports as well as multiple wildcard imports from the same module.  To enable the new behavior pass the following object as an option to the rule in your tslint.json: `{ "allow-namespace-imports": true }`

#### CHANGELOG.md entry:

[new-rule-option] added `allow-namespace-imports ` option to `no-duplicate-imports` rule

palantirtech commented:
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @EdwardDrapkin! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

JoshuaKGoldberg commented on src/rules/noDuplicateImportsRule.ts:
```diff
@@ -41,21 +47,43 @@ export class Rule extends Lint.Rules.AbstractRule {
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithFunction(sourceFile, walk);
+        return this.applyWithFunction(sourceFile, (ctx: Lint.WalkContext<void>) =>
```
This looks like it works, but to simplify a bit, you can pass an options object as a third argument to `applyWithFunction`. That'll allow you to access it on `ctx.options` instead of passing a new parameter.

JoshuaKGoldberg commented on src/rules/noDuplicateImportsRule.ts:
```diff
@@ -29,9 +31,13 @@ export class Rule extends Lint.Rules.AbstractRule {
+        optionsDescription: Lint.Utils.dedent`
+            "${OPTION_ALLOW_SEPARATE_NAMESPACE_IMPORTS}" allows you to import namespaces on separate lines.`,
+        options: {
+            type: "string",
```
Although TSLint has traditionally had rule arguments provided as strings, that's starting to be a bit of a pain when they become more complex. Let's have the rule take in an object:

```ts
[true, {
    "allow-namespace-imports": true
}]
```

`memberOrderingRule.ts` is a good example of this.

EdwardDrapkin commented:
Thanks for the feedback.

I believe I have addressed all your issues, but for brevity, here are my comments on your comments.

1. I updated the call to pass options per your suggestion.  Didn't know this was possible, thanks for the help.

2. I figured out how to write tests and added a test case for this option because the option is a bit more complex.  I believe I covered all of the possible permutations of error/not error when the rule is enabled.

3. I updated the option to be an object with the format `{ "allow-namespace-imports": boolean }`.

4.  The rule now tracks named imports separately from namespace imports.  The benefit is now that multiple wildcard imports will error out alongside multiple named imports.  You also get a separate, special error message specifically for multiple wildcard imports.

Thanks for your help!!

EdwardDrapkin commented:
Looks like all the test passed this time 🎉

JoshuaKGoldberg commented:
Yes, the changelog entry is a bit too long haha. I'll go ahead and edit it.

huyz commented:
Is there a reason why this option isn't the default?

JoshuaKGoldberg commented:
@huyz it happened not to be in this PR, but that's a good idea. Perhaps it should - would you mind filing a new issue?

EdwardDrapkin commented:
> Is there a reason why this option isn't the default?

I didn't want to presume to change default behavior is the only reasoning I had to make it opt-in instead of opt-out.
